plugin-grace-hopper/src/probe.rs

```diff
@@ -100,14 +102,45 @@ pub fn read_power_value(buffer: &mut String, file: &mut File) -> Result<u64, any
     Ok(power_consumption)
 }
 
+/// Compute an energy from a power. Using as time the time elapsed between
+/// `last_timestamp` and the current timestamp `timestamp`.
+///
+/// This function first computes as `u64` the current Timestamp as ns. Then using the
+/// `CounterDiff` structure it compute the time elapsed since the last measurement.
+/// Finally it compute the energy using the formula: Energy(J) = Power(W) * Time(s)
+///
+/// Returns the computed energy on success or 0.0 for the first time
+pub fn compute_energy(
+    power: u64,
+    last_timestamp: &mut CounterDiff,
+    timestamp: Timestamp,
+) -> Result<f64, anyhow::Error> {
+    let (second, nanosec) = timestamp.to_unix_timestamp();
+    let final_value = second * 1_000_000_000 + nanosec as u64;
+    let time_elapsed_opt = match last_timestamp.update(final_value) {
+        CounterDiffUpdate::FirstTime => None,
+        CounterDiffUpdate::Difference(diff) | CounterDiffUpdate::CorrectedDifference(diff) => Some(diff),
+    };
+    if let Some(time_elapsed_ns) = time_elapsed_opt {
+        let time_in_seconds_u64 = time_elapsed_ns / 1_000_000_000;
+        let time_in_seconds = time_in_seconds_u64 as f64;
+        Ok(power as f64 * time_in_seconds)
+    } else {
+        Ok(power as f64)
```

Comment:
in case counterdiff is None (here it's the case) I think returning power as is since is wrong because we still want an energy, so I would rather return an Ok Option None